### PR TITLE
Compute Good_turing estimate

### PR DESCRIPTION
**Pull request by dliyanage**

### CODE REVIEW

File: src/afl-fuzz-run.c [hunk 1]
```diff
@@ -589,7 +589,7 @@ u8 calibrate_case(afl_state_t *afl, struct queue_entry *q, u8 *use_mem,
 
     stop_us = get_cur_time_us();
     diff_us = stop_us - start_us;
-    if (unlikely(!diff_us)) { ++diff_us; }
+    if (unlikely(!diff_us)) { diff_us = 1; }
```
Comment (mboehme) [inline src/afl-fuzz-run.c]: Revert.

File: src/afl-fuzz-bitmap.c [hunk 2]
```diff
@@ -483,9 +483,13 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
-    if (likely(afl->n_fuzz[cksum % N_FUZZ_SIZE] < 0xFFFFFFFF))
-      afl->n_fuzz[cksum % N_FUZZ_SIZE]++;
+    if (afl->n_fuzz[cksum % N_FUZZ_SIZE] < 0xFFFFFFFF) {
 
+      u32 entry = cksum % N_FUZZ_SIZE;
+      afl->n_fuzz[entry]++;
+      afl->n_fuzz_stats[entry]++;
+
+    }
   }
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Revert (Redundant).

File: src/afl-fuzz-bitmap.c [hunk 3]
```diff
@@ -483,9 +483,13 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
     cksum = hash64(afl->fsrv.trace_bits, afl->fsrv.map_size, HASH_CONST);
 
     /* Saturated increment */
-    if (likely(afl->n_fuzz[cksum % N_FUZZ_SIZE] < 0xFFFFFFFF))
-      afl->n_fuzz[cksum % N_FUZZ_SIZE]++;
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Here we are tracking hit counts. We are only interested in the number of singletons, whether reset or not.
So, let's define a global variable called `singletons`, and 
* every time `afl->nfuzz[cksum % N_FUZZ_SIZE] == 0` before the hit count increment, we increment `singletons`.
* every time `afl->nfuzz[cksum % N_FUZZ_SIZE] == 1` before the hit count increment, we decrement `singletons`.

This optimizes the computation of singletons. You don't need to go through the entire array.

File: src/afl-fuzz-stats.c [hunk 4]
```diff
@@ -283,6 +284,28 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
...
   }
 
+  if (afl->schedule >= FAST && afl->schedule <= RARE) {
+
+    u64 singletons = 0;
+    for (u32 i = 0; i < N_FUZZ_SIZE; ++i) {
+
+      if (afl->n_fuzz_stats[i] == 1) { ++singletons; }
+
+    }
```
Comment (mboehme) [inline src/afl-fuzz-stats.c]: Now redundant.

File: src/afl-fuzz-bitmap.c [hunk 5]
```diff
@@ -526,13 +530,20 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
-#endif                                                    /* ^!SIMPLE_FILES */
+#endif                                                    
+    /* ^!SIMPLE_FILES */
     fd = open(queue_fn, O_WRONLY | O_CREAT | O_EXCL, DEFAULT_PERMISSION);
     if (unlikely(fd < 0)) { PFATAL("Unable to create '%s'", queue_fn); }
     ck_write(fd, mem, len, queue_fn);
     close(fd);
     add_to_queue(afl, queue_fn, len, 0);
 
+    if (afl->queued_paths % 10 == 0) {
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Make constant a "variable". For instance, see other constants in afl-fuzz.h.

File: src/afl-fuzz-bitmap.c [hunk 6]
```diff
@@ -526,13 +530,20 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
+
+      memset(afl->n_fuzz_stats, 0, N_FUZZ_SIZE * sizeof(u32));
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Now, you can set `singletons = 0`.

File: src/afl-fuzz-stats.c [hunk 7]
```diff
@@ -283,6 +284,28 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
...
+
+    // Compute the Good-Turing estimates
+    if (afl->fsrv.total_execs > 0) {
+
+        u64 gt = (singletons /afl->fsrv.total_execs);
```
Comment (mboehme) [inline src/afl-fuzz-stats.c]: Here you redefine gt in a local scope. As soon as you leave the if statement, the value is forgotten.
Instead, remove the variable type (u64) to let this value be assigned to src/afl-fuzz-stats.c#L254.

File: src/afl-fuzz-stats.c [hunk 8]
```diff
@@ -297,6 +320,7 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
       "cycles_wo_finds   : %llu\n"
       "time_wo_finds     : %llu\n"
       "execs_done        : %llu\n"
+      "good-turing       : %llu\n"
```
Comment (mboehme) [inline src/afl-fuzz-stats.c]: Also print it to plot_data.

Comment (mboehme) [reply src/afl-fuzz-stats.c]: Line 490.

**Review by mboehme — COMMENTED** [review]
As discussed.

File: include/afl-fuzz.h [hunk 9]
```diff
@@ -552,7 +559,7 @@ typedef struct afl_state {
   u8 *var_bytes;                        /* Bytes that appear to be variable */
 
 #define N_FUZZ_SIZE (1 << 21)
```
Comment (mboehme) [inline include/afl-fuzz.h:563]:
```C
#define RESET_PARAM  1 /* 10 */
```

File: include/afl-fuzz.h [hunk 10]
```diff
@@ -541,6 +541,13 @@ typedef struct afl_state {
       old_seed_selection,               /* use vanilla afl seed selection   */
       reinit_table;                     /* reinit the queue weight table    */
 
+  u64 singletons,                        /* Number of singletons */
+      reset_param,                      /* The number of seeds to add before resetting the hit counts */
```
Comment (mboehme) [inline include/afl-fuzz.h]: Substitute with constant (see below).

File: src/afl-fuzz-bitmap.c [hunk 11]
```diff
@@ -455,6 +455,8 @@ void write_crash_readme(afl_state_t *afl) {
 u8 __attribute__((hot))
 save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
 
+  afl->reset_param = 1;
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Remove (once you have that as a constant)

File: src/afl-fuzz-bitmap.c [hunk 12]
```diff
@@ -482,9 +484,40 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
+
+      }
+
+    afl->singletons_reset = 0;
+      for (u32 i = 0; i < afl->queued_items; i++) {
+
+        struct queue_entry *q = afl->queue_buf[i];
+        if (afl->n_fuzz_reset[q->n_fuzz_entry] == 1) { ++afl->singletons_reset; }
+
+      }
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Comment this.

File: src/afl-fuzz-bitmap.c [hunk 13]
```diff
@@ -482,9 +484,40 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
 
+    // Update the nunber of singletons and reset-singletons
+    /*if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 0 ){
+      afl->singletons--;
+      afl->singletons_reset--;
+    }
+    if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 1 ){
+      afl->singletons++;
+      afl->singletons_reset++;
+    }
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]: Uncomment this, but set `afl->singletons = 0` whenever everything else is set up, too.

File: src/afl-fuzz-bitmap.c [hunk 14]
```diff
@@ -482,9 +484,40 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
 
     cksum = hash64(afl->fsrv.trace_bits, afl->fsrv.map_size, HASH_CONST);
 
+    // Update the nunber of singletons and reset-singletons
+    /*if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 0 ){
+      afl->singletons--;
+      afl->singletons_reset--;
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]:
```C
if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 0 ) afl->singletons++;
if (afl->n_fuzz_reset[cksum % N_FUZZ_SIZE] == 0 ) afl->singletons_reset++;
```

File: src/afl-fuzz-bitmap.c [hunk 15]
```diff
@@ -482,9 +484,40 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
+    }
+    if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 1 ){
+      afl->singletons++;
+      afl->singletons_reset++;
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]:
```C
if (afl->n_fuzz[cksum % N_FUZZ_SIZE] == 1) afl->singletons--;
if (afl->n_fuzz_reset[cksum % N_FUZZ_SIZE] == 1) afl->singletons_reset--;
```

File: src/afl-fuzz-bitmap.c [hunk 16]
```diff
@@ -526,13 +559,21 @@ save_if_interesting(afl_state_t *afl, void *mem, u32 len, u8 fault) {
...
-#endif                                                    /* ^!SIMPLE_FILES */
+#endif                                                    
+    /* ^!SIMPLE_FILES */
     fd = open(queue_fn, O_WRONLY | O_CREAT | O_EXCL, DEFAULT_PERMISSION);
     if (unlikely(fd < 0)) { PFATAL("Unable to create '%s'", queue_fn); }
     ck_write(fd, mem, len, queue_fn);
     close(fd);
     add_to_queue(afl, queue_fn, len, 0);
 
+    if (afl->queued_favored % afl->reset_param == 0) {
```
Comment (mboehme) [inline src/afl-fuzz-bitmap.c]:
```C
if ((afl->queued_items % RESET_PARAM) == 0) {
  memset(afl->n_fuzz_reset, 0, N_FUZZ_SIZE * sizeof(u32));
  afl->singletons_reset = 0;
}
```

File: src/afl-fuzz.c [hunk 17]
```diff
@@ -1590,6 +1590,7 @@ int main(int argc, char **argv_orig, char **envp) {
   if (afl->schedule >= FAST && afl->schedule <= RARE) {
 
     afl->n_fuzz = ck_alloc(N_FUZZ_SIZE * sizeof(u32));
+    afl->n_fuzz_reset = ck_alloc(N_FUZZ_SIZE * sizeof(u32));
```
Comment (mboehme) [inline src/afl-fuzz.c]:
```C
afl->singletons = 0;
afl->singletons_reset = 0;
```

File: src/afl-fuzz-stats.c [hunk 18]
```diff
@@ -283,6 +283,22 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
 
   }
 
+  if (afl->schedule >= FAST && afl->schedule <= RARE) {
+    // Compute the Good-Turing estimates
+    if (afl->fsrv.total_execs > 0) {
+
+        afl->gt = (afl->singletons /afl->fsrv.total_execs);
+        afl->gt_reset = (afl->singletons_reset / afl->fsrv.total_execs);
```
Comment (mboehme) [inline src/afl-fuzz-stats.c]:
```C
afl->gt = ((double) afl->singletons) / afl->fsrv.total_execs;
afl->gt_reset = ((double) afl->singletons_reset) / afl->fsrv.total_execs;
```

**Review by mboehme — CHANGES_REQUESTED** [review]
Thanks! Several suggested changes below.

**Comment by dliyanage** [comment]
All changes implemented. Both general and 1-reset Good-Turing estimators now computed well. See the below outputs.
Screenshot 2023-09-11 at 10 28 36 am
Screenshot 2023-09-11 at 10 28 52 am

**Comment by mboehme** [comment]
Nice! 

Just wanted to check something. Instead of the floating point result, can you print the number of singletons (and reset singletons)? This way, we measure singletons, I suspect f1 to increase proportionally to the total number of generated inputs (i.e., almost everything is a singleton).

**Review by mboehme — APPROVED** [review]

**Pull request merged**